File: sdk/ts/src/schemes/xmss.ts
/** DKKW25 generalized XMSS. Signing state belongs to the caller. */
import { Address, TransactionInstruction } from "@solana/web3.js";
import { Scheme, sha256 } from "../scheme.js";
import { createInitializeInstruction } from "../instructions.js";

// Mirror solana-winternitz's fixed height-8 instance.
export const XMSS_PUBKEY_LEN = 41;
export const XMSS_SIGNATURE_LEN = 1037;

export const XMSS: Scheme = {
  programId: new Address("7qCyy3NJQDMctSDiM4DxNjNR6TyasouyyRTBREhcXdsE"),
  signatureLen: XMSS_SIGNATURE_LEN,
  identityLen: 32,
  storedIdentityLen: 32 + XMSS_PUBKEY_LEN,
};

/** Derive the permanent account identity. Do not recompute it on rotation. */
export function xmssIdentity(initialPublicKey: Uint8Array): Uint8Array {
  if (initialPublicKey.length !== XMSS_PUBKEY_LEN) {
    throw new Error(`XMSS public key must be ${XMSS_PUBKEY_LEN} bytes`);
  }
  return sha256(initialPublicKey);
}

/** Initialize a stable account with its first signing key. */
export function createInitializeXmss(
  payer: Address,
  publicKey: Uint8Array
): TransactionInstruction {
  return createInitializeInstruction(payer, XMSS, xmssIdentity(publicKey), publicKey);
}

File: crates/core/src/schemes/xmss.rs
//! DKKW25 generalized XMSS. Sign through `solana_winternitz::Signer`:
//! each key permits 256 signing attempts, including failed salt sampling.

use solana_address::{address, Address};
use solana_instruction::Instruction;
use solana_winternitz::{xmss, PUBLIC_KEY_LENGTH};

use crate::{instructions::create_initialize_instruction, scheme::Scheme};

pub const XMSS_PUBKEY_LEN: usize = PUBLIC_KEY_LENGTH;
pub const XMSS_SIGNATURE_LEN: usize = xmss::SIGNATURE_LENGTH;

/// The public key is stored verbatim; the PDA seed is its SHA-256 hash.
pub const XMSS: Scheme = Scheme {
    program_id: address!("7qCyy3NJQDMctSDiM4DxNjNR6TyasouyyRTBREhcXdsE"),
    signature_len: XMSS_SIGNATURE_LEN,
    identity_len: XMSS_PUBKEY_LEN,
    stored_identity_len: XMSS_PUBKEY_LEN,
};

/// Initialize an account with an immutable XMSS public key.
pub fn create_initialize_xmss(payer: &Address, public_key: &[u8; XMSS_PUBKEY_LEN]) -> Instruction {
    create_initialize_instruction(payer, &XMSS, public_key, public_key)
}
